lib/Hakyll/Core/Compiler/Internal.hs

```diff
@@ -166,30 +182,29 @@ instance MonadMetadata Compiler where
 --------------------------------------------------------------------------------
 instance MonadError [String] Compiler where
   throwError = compilerThrow
-  catchError = (. matchErr) . compilerCatch
-    where
-      matchErr f Logger.Error es = f es
-      matchErr f _            _  = f []
+  catchError c = compilerCatch c . (. getReason)
 
 
 --------------------------------------------------------------------------------
 runCompiler :: Compiler a -> CompilerRead -> IO (CompilerResult a)
 runCompiler compiler read' = handle handler $ unCompiler compiler read'
   where
     handler :: SomeException -> IO (CompilerResult a)
-    handler e = return $ CompilerError Logger.Error [show e]
+    handler e = return $ CompilerError $ CompilationFailure [show e]
 
 
 --------------------------------------------------------------------------------
 instance Alternative Compiler where
-    empty   = compilerError Logger.Debug []
-    x <|> y = compilerCatch x $ \vx exs -> compilerCatch y $ \vy eys ->
-        case vx `compare` vy of
-            LT -> log eys >> compilerError vx exs
-            EQ ->            compilerError vx (exs ++ eys)
-            GT -> log exs >> compilerError vy eys
+    empty   = compilerMissing []
+    x <|> y = x `compilerCatch` (\rx -> y `compilerCatch` (\ry ->
+        case (rx, ry) of
+          (CompilationFailure xs,  CompilationFailure ys)  -> compilerThrow $ xs ++ ys
+          (CompilationFailure xs,  NoCompilationResult ys) -> debug ys >> compilerThrow xs
+          (NoCompilationResult xs, CompilationFailure ys)  -> debug xs >> compilerThrow ys
+          (NoCompilationResult xs, NoCompilationResult ys) -> compilerMissing $ xs ++ ys
```

Comment:
I could have made `Reason` a `Semigroup` or `Monoid` instance for this to make the code even cleaner, but that doesn't allow to put the discarded messages in the `debug`.